### DISCUSSION TITLE
Are pretrained models available?

### DISCUSSION THREAD

**Discussion by mattragoza** [RMeli/gnina-torch#40] [Q&A] [answered]
I would love to use this module within liGAN instead of depending on the full gnina platform. Have the caffe model weights been converted to pytorch somewhere publicly available?

**Comment by RMeli** [comment]
Unfortunately, I never converted the original Caffe models weights to PyTorch. The `examples` folder contains some scripts to reproduce part of the results of your and Paul's work, but are meant mostly for testing.

However, I would be very interested in doing so (or incorporate PRs that will allow to do that). It would be a great feature, and would serve as double check that everything works as expected. Is there any recommended way of doing this Caffe to PyTorch conversion?

PS: For easier integration, I could push the package to PyPI and conda-forge, if it is of interest.

**Comment by drewnutt** [comment]
I actually have these weights files for all of the built-in models easily available. 

The built-in model weights were pulled directly from the source code of GNINA and converted to a `.caffemodel`. Then https://github.com/vadimkantorov/caffemodel2pytorch was utilized to convert the `.caffemodel` files to `.pt` for PyTorch. 

I have created PR #34 to add these weights files.

**Reply by Kerro-junior** [reply]
@drewnutt  Hi, would you please tell me where can I find the caffemodel for built-in `*default2018.model`? I have generated types file ready for training, as format \<label> \<pk> \<rmsd> \<rec> \<lig>:
```
0 1.9966 3.0868 pdb2019_refi_train_gninatypes/185l/185l_rec.gninatypes first_pdbbind_v2019_docked_gninatypes/185l_docked_0.gninatypes
1 2.8809 1.3182 pdb2019_refi_train_gninatypes/185l/185l_rec.gninatypes first_pdbbind_v2019_docked_gninatypes/185l_docked_1.gninatypes
```

I've searched for days but couldn't find their **.caffemodel weights** to use in train.py to use it as a pre-trained task, when I searched 'caffemodel' there is just the following corresponding results:
![image](https://github.com/RMeli/gnina-torch/assets/139468131/592256fb-5245-49a6-8cdd-40c2042776a1)
![image](https://github.com/RMeli/gnina-torch/assets/139468131/ef0181b6-6d15-4c14-9e3f-c0c9275b0b7a)

**Reply by RMeli** [reply]
@Kerro-junior this question is related to GNINA, not `gnina-torch`. There are no Caffe weights in this repository, only their corresponding PyTorch version.

You can find the model definitions for GNINA in the GNINA repo here: https://github.com/gnina/gnina/tree/master/gninasrc/lib/models

And the mode weights are here: https://github.com/gnina/gnina/tree/master/gninasrc/lib/models/weights

**Reply by drewnutt** [reply]
@Kerro-junior I have uploaded the weights files for all of the default2018 models to [gnina/models/crossdocked_paper](https://github.com/gnina/models/tree/master/crossdocked_paper) with [(gnina/models/pull/31)](https://github.com/gnina/models/pull/31)

**Comment by dkoes** [comment]
I'll add that there is now a `dumpmodel` binary built when you build gnina that will export the built-in caffe model weights (but not that actual caffe model file as you can get this easily from the source code).

**Comment by RMeli** [comment]
That's fantastic, thanks @drewnutt ! I'll incorporate PR #34 and start adding the functionality to use them easily (including model ensemble, which is currently not implemented).

**Comment by mattragoza** [comment]
Awesome! Thanks, @drewnutt.

**Comment by RMeli** [comment]
I opened #35 as a follow-up. I'll link relevant PRs there. Please feel free to comment with specific needs and missing features from your point of view.

**Comment by RMeli** [comment]
@mattragoza you can now easily load the pre-trained `default2017` and `*default2018*` model as follows:
```python
from gninatorch.gnina import load_gnina_model

model = load_gnina_model(MODEL_NAME)
```
where `MODEL_NAME` corresponds to name of the `.pt` files in `gninatorch/weights` (without the `.pt` extension); this is essentially equivalent to GNINA's `--cnn` option (but without model ensembles, which will follow).

The `dense` model is not yet supported because I haven't managed to obtain the same output as GNINA.

**Comment by RMeli** [comment]
An ensemble of models is now also available:
```python
from gninatorch.gnina import load_gnina_models

ensemble = load_gnina_models([MODEL_NAME_1, MODEL_NAME_2, ...])
````
The ensemble of models returns `log_CNNscore`, `CNNaffinity` and `CNNvariance`.

Still no `dense` model for the time being (I suspect that there might be an issue with `BatchNorm` layer weights, but I'm still investigating), so it is not currently possible to reproduce GNINA's `default` model (since it contains two `dense` models). However, `*_default2018_ensemble` work as expected.

**Comment by RMeli** [comment, marked as answer]
@mattragoza Thanks once again to @drewnutt , who manually converted the weights of the `dense` model (see #43), the `dense_` models and the `dense_ensemble` are now supported. The same goes for GNINA's `default` model ensemble. 

See PR #44 for details. Related PRs: #38, #36, #34.

You can now easily load GNINA's pre-trained models as follows:
```python
from gninatorch import gnina

model, ensemble = setup_gnina_model("MODEL")
```
where `MODEL` corresponds to the string you would use as `--cnn` argument in GNINA, while `ensemble` is a boolean flag that tells you if `model` is a single model (`ensemble == False`) or if it is an ensemble of models (`ensemble == True`).

**Reply by RMeli** [reply]
Now `pip install gninatorch` should do the trick, and by default will pull the dependencies needed to load and use GNINA models.
